xarray/core/dataset.py

```diff
@@ -1110,16 +1124,66 @@ def relevant_keys(mapping):
                 # dim already exists
                 raise ValueError('Existing dimensions are not valid choices '
                                  'for the dim argument in sel_points')
+            # replace the array dim
+            dim_coord = dim.copy()
+            dim = dim.name
 
         if not utils.is_scalar(dim) and not isinstance(dim, DataArray):
-            dim = as_variable(dim, name='points')
+            name = 'points' if not hasattr(dim, 'name') else dim.name
+            dim_coord = as_variable(dim, name=name)
+            dim = name
 
-        # TODO: This would be sped up with vectorized indexing. This will
-        # require dask to support pointwise indexing as well.
-        return concat([self.isel(**d) for d in
-                       [dict(zip(keys, inds)) for inds in
-                        zip(*[v for k, v in indexers])]],
-                      dim=dim, coords=coords, data_vars=data_vars)
+        reordered = self.transpose(*(list(indexer_dims) + list(non_indexed_dims)))
+
+        variables = OrderedDict()
+        sel_coords = OrderedDict()
+
+        for name, var in reordered.variables.items():
+            if name in indexers_dict or any(d in indexer_dims for d in var.dims):
+                # if the var is an indexer or if it depends on an indexed dim, slice it
+                slc = [indexers_dict[k] if k in indexers_dict else slice(None) for k in var.dims]
+                var_dims = [dim] + [d for d in var.dims if d in non_indexed_dims]
+                selection = take(var, tuple(slc))
+
+                # Add the var to variables or coords accordingly
+                if name in coords:
+                    sel_coords[name] = (var_dims, selection)
```

Comment:
Also note: recreating variables with a tuple drops `attrs`. It would be better to write `(var_dims, selection, var.attrs)` or better yet `type(var)(var_dims, selection, var.attrs)` to be explicit about constructing variable objects (you would need to do this if you switch to use `_replace_vars_and_dims`).